### ISSUE TITLE
urllib.request.urlopen does not handle non-ASCII characters

### ISSUE THREAD

**Issue by 526a9556-0b1e-466b-8760-d36ea509066e** [python/cpython#48241]
BPO | [3991](https://bugs.python.org/issue3991)
--- | :---
Nosy | @orsenthil, @vstinner, @devdanzin, @ezio-melotti, @abadger, @bitdancer, @vadmium, @remilapeyre
Dependencies | <li>bpo-9679: unicode DNS names in urllib, urlopen</li>
Files | <li>[non_ascii_path.diff](https://bugs.python.org/file12986/non_ascii_path.diff "Uploaded as text/plain at 2009-02-08.21:50:18 by @devdanzin"): Calls quote() on the request path if path.encode('ascii') fails</li><li>[issue3991.diff](https://bugs.python.org/file29194/issue3991.diff "Uploaded as text/plain at 2013-02-23.17:58:19 by thezulk")</li><li>[issue3991_2017-01-27.diff](https://bugs.python.org/file46440/issue3991_2017-01-27.diff "Uploaded as text/plain at 2017-01-27.21:58:29 by thezulk"): Diff against 3.7.0a0</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2008-09-28.18:47:16.056>
labels = ['extension-modules', '3.8', 'type-bug', '3.7', 'easy']
title = 'urllib.request.urlopen does not handle non-ASCII characters'
updated_at = <Date 2019-02-20.20:02:12.619>
user = 'https://github.com/abadger'
```

bugs.python.org fields:
```python
activity = <Date 2019-02-20.20:02:12.619>
actor = 'remi.lapeyre'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Extension Modules']
creation = <Date 2008-09-28.18:47:16.056>
creator = 'a.badger'
dependencies = ['9679']
files = ['12986', '29194', '46440']
hgrepos = []
issue_num = 3991
keywords = ['patch', 'easy']
message_count = 14.0
messages = ['73982', '74046', '74053', '74085', '74088', '74110', '74117', '81423', '182785', '219325', '285721', '286386', '286423', '286444']
nosy_count = 11.0
nosy_names = ['janssen', 'orsenthil', 'vstinner', 'ajaksu2', 'ezio.melotti', 'a.badger', 'r.david.murray', 'martin.panter', 'thezulk', 'Graham.Oliver', 'remi.lapeyre']
pr_nums = []
priority = 'normal'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = 'behavior'
url = 'https://bugs.python.org/issue3991'
versions = ['Python 3.4', 'Python 3.5', 'Python 3.6', 'Python 3.7', 'Python 3.8']
```

</p></details>

### Linked PRs
* gh-103855
* gh-103891

**Comment by mblahay** [comment]
I am suggesting that this be changed to a documentation issue.

**Comment by ambv** [comment]
Agreed. The domain name would have to be encoded in [Punycode](https://en.wikipedia.org/wiki/Punycode), while the rest of the path uses percent-encoding, which depends on the byte representation that the server expects. It's unclear in general to which bytes, say, "ł" should translate on the server side.

Documenting that we're not doing automatic encoding of either domains or path elements is the most reasonable way forward given that it's been 15 years since this issue got created.

**Comment by mblahay** [comment]
@ambv Is this enough?

Open the URL *url*, which can be either a string containing a valid, properly encoded, URL or a Request object.

I am attempting to avoid specifically calling out the encodings to use as it may be difficult to communicate succinctly and they may change over time.

**Comment by ambv** [comment]
Yes, looks good. Let's see it on a PR.

**Comment by erlend-aasland** [comment]
Resolved by:

- gh-103855
- gh-103891

**Comment by merwok** [comment]
Closes #64758 (duplicate)
